src/wnaf.rs

```diff
@@ -349,46 +351,91 @@ impl<B, S: AsMut<Vec<i64>>> Wnaf<usize, B, S> {
     }
 }
 
-/// A window size that is fixed in the type system.
-pub struct FixedWindow<const WINDOW_SIZE: usize>;
+/// A "w-ary non-adjacent form" scalar, that uses precomputation to improve the speed of
+/// scalar multiplication.
+///
+/// # Examples
+///
+/// See [`WnafBase`] for usage examples.
+#[derive(Clone, Debug)]
+pub struct WnafScalar<F: PrimeField, const WINDOW_SIZE: usize> {
+    wnaf: Vec<i64>,
+    field: PhantomData<F>,
+}
 
-impl<G: WnafGroup, const WINDOW_SIZE: usize>
-    Wnaf<FixedWindow<WINDOW_SIZE>, PhantomData<G>, Vec<i64>>
-{
-    /// Given a scalar, compute its wNAF representation with the specified `WINDOW_SIZE`
-    /// and return a `Wnaf` object that can perform exponentiations with `Wnaf::base(..)`.
-    pub fn scalar(scalar: &<G as Group>::Scalar) -> Self {
+impl<F: PrimeField, const WINDOW_SIZE: usize> WnafScalar<F, WINDOW_SIZE> {
+    /// Computes the w-NAF representation of the given scalar with the specified
+    /// `WINDOW_SIZE`.
+    pub fn new(scalar: &F) -> Self {
         let mut wnaf = vec![];
 
         // Compute the wNAF form of the scalar.
         wnaf_form(&mut wnaf, scalar.to_repr(), WINDOW_SIZE);
 
-        Wnaf {
-            base: PhantomData::default(),
-            scalar: wnaf,
-            window_size: FixedWindow,
+        WnafScalar {
+            wnaf,
+            field: PhantomData::default(),
         }
     }
 }
 
-impl<G: WnafGroup, const WINDOW_SIZE: usize> Wnaf<FixedWindow<WINDOW_SIZE>, Vec<G>, ()> {
-    /// Given a base, computes a window table with the specified `WINDOW_SIZE` and returns
-    /// a `Wnaf` object that can perform exponentiations with `Wnaf::scalar(..)`.
-    pub fn base(base: G) -> Self {
+/// A fixed window table for a group element, precomputed to improve the speed of scalar
+/// multiplication.
+///
+/// This struct is designed for usage patterns that have long-term cached bases and/or
+/// scalars, or [Cartesian products] of bases and scalars. The [`Wnaf`] API enables one or
+/// the other to be cached, but requires either the base window tables or the scalar w-NAF
+/// forms to be computed repeatedly on-the-fly, which can become a significant performance
```

Comment:
```suggestion
/// forms to be computed repeatedly on the fly, which can become a significant performance
```
"on the fly" is not a compound adjective here.